oak_crypto/src/noise_nk/rustcrypto.rs

```diff
@@ -0,0 +1,212 @@
+
+use aes_gcm::{AeadInPlace, KeyInit};
+use p256::ecdsa::signature::Signer;
+use pkcs8::{DecodePrivateKey, EncodePrivateKey};
+use primeorder::{
+    elliptic_curve::{
+        ops::{Mul, MulByGenerator},
+        sec1::{FromEncodedPoint, ToEncodedPoint},
+    },
+    Field, PrimeField,
+};
+use sha2::Digest;
+
+pub fn rand_bytes(_output: &mut [u8]) {
+    panic!("unimplemented");
+}
+
+/// Perform the HKDF operation from https://datatracker.ietf.org/doc/html/rfc5869
+pub fn hkdf_sha256(ikm: &[u8], salt: &[u8], info: &[u8], output: &mut [u8]) -> Result<(), ()> {
+    hkdf::Hkdf::<sha2::Sha256>::new(Some(salt), ikm).expand(info, output).map_err(|_| ())
+}
+
+pub fn aes_256_gcm_seal_in_place(
+    key: &[u8; 32],
+    nonce: &[u8; NONCE_LEN],
+    aad: &[u8],
+    plaintext: &mut Vec<u8>,
+) {
+    aes_gcm::Aes256Gcm::new_from_slice(key.as_slice())
+        .unwrap()
+        .encrypt_in_place(nonce.into(), aad, plaintext)
+        .unwrap();
+}
+
+pub fn aes_256_gcm_open_in_place(
+    key: &[u8; 32],
+    nonce: &[u8; NONCE_LEN],
+    aad: &[u8],
+    mut ciphertext: Vec<u8>,
+) -> Result<Vec<u8>, ()> {
+    aes_gcm::Aes256Gcm::new_from_slice(key.as_slice())
+        .unwrap()
+        .decrypt_in_place(nonce.into(), aad, &mut ciphertext)
+        .map_err(|_| ())?;
+    Ok(ciphertext)
+}
+
+pub fn sha256(input: &[u8]) -> [u8; SHA256_OUTPUT_LEN] {
+    let mut ctx = sha2::Sha256::new();
+    ctx.update(input);
+    ctx.finalize().into()
+}
+
+/// Compute the SHA-256 hash of the concatenation of two inputs.
+pub fn sha256_two_part(input1: &[u8], input2: &[u8]) -> [u8; SHA256_OUTPUT_LEN] {
+    let mut ctx = sha2::Sha256::new();
+    ctx.update(input1);
+    ctx.update(input2);
+    ctx.finalize().into()
+}
+
+pub struct P256Scalar {
+    v: p256::Scalar,
+}
+
+impl P256Scalar {
+    pub fn generate() -> P256Scalar {
+        let mut ret = [0u8; P256_SCALAR_LENGTH];
+        // Warning: not very random.
+        ret[0] = 1;
+        P256Scalar { v: p256::Scalar::from_repr(ret.into()).unwrap() }
+    }
+
+    pub fn compute_public_key(&self) -> [u8; P256_X962_LENGTH] {
+        p256::ProjectivePoint::mul_by_generator(&self.v)
+            .to_encoded_point(false)
+            .as_bytes()
+            .try_into()
+            .unwrap()
+    }
+
+    pub fn bytes(&self) -> [u8; P256_SCALAR_LENGTH] {
+        self.v.to_repr().as_slice().try_into().unwrap()
+    }
+}
+
+impl TryFrom<&[u8]> for P256Scalar {
+    type Error = ();
+
+    fn try_from(bytes: &[u8]) -> Result<Self, ()> {
+        let array: [u8; P256_SCALAR_LENGTH] = bytes.try_into().map_err(|_| ())?;
+        (&array).try_into()
+    }
+}
+
+impl TryFrom<&[u8; P256_SCALAR_LENGTH]> for P256Scalar {
+    type Error = ();
+
+    fn try_from(bytes: &[u8; P256_SCALAR_LENGTH]) -> Result<Self, ()> {
+        let scalar = p256::Scalar::from_repr((*bytes).into());
+        if !bool::from(scalar.is_some()) {
+            return Err(());
+        }
+        let scalar = scalar.unwrap();
```

Comment:
This can be done with regular pattern matching without unwraps.

Reply:
That would be better if scalar were an Option<T>, but it is a CtOption<T>.  IIUC, the enum values for CtOption is private on purpose, so we don't branch on a secret T, which would violate constaint-time rules.

Reply:
Oh interesting, thanks! But I am a bit confused -- are we not branching on whether this was `some` or not here?